File: nodejs/paillier.js
var bignum = require('bignum');

const letters = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q",
    "r", "s", "t", "u", "v", "w", "x", "y", "z", ",", ".", "!", "?", ' '
];
const numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,
    14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27,
    28, 29, 30, 31
];
const keyLength = 10;



function generatePaillierKeyPair() {
    var p = bignum.prime(keyLength / 2);
    var q = bignum.prime(keyLength / 2);
    //if I put p q bigger, it takes too much time to decrypt
    p = 29;
    q = 479;
    var compr = gcd(p * q, (p - 1) * (q - 1));
    /*console.log("gcd(p*q, (p-1)*(q-1)) = " + compr);
    console.log("p: " + p + ", q: " + q);*/

    var n = bignum(p * q);
    //console.log("n= " + n);

    var lambda = lcm(p - 1, q - 1);

    var alpha = bignum(rand(0, n.toNumber()));
    var beta = bignum(rand(0, n.toNumber()));
    var alphan = alpha.mul(n);
    var alphan1 = alpha.add(1);
    var betaN = beta.pow(n);
    var ab = alphan1.mul(betaN);
    var n2 = n.pow(2);
    var g = ab.mod(n2);
    //in some Paillier implementations use this:
    //var g = n.add(1);

    var pubK = {
        n: n,
        g: g
    };

    //var lFunc = L(bignum(g.pow(lambda)).mod(n*n).toNumber(), n);
    var Glambda = g.pow(lambda);
    var u = Glambda.mod(n2);
    var lFunc = L(u, n);
    var mu = lFunc.invertm(n);


    var privK = {
        lambda: bignum(lambda),
        mu: mu
    };
    return ({
        "pubK": pubK,
        "privK": privK
    });
}



function encrypt(m, pubK) {
    chars = m.split('');
    var r = [];
    for (var i = 0; i < chars.length; i++) {
        numb = getNbyL(chars[i]);
        r.push(encryptNum(numb, pubK));
    }
    return r;
}

function decrypt(c, pubK, privK) {
    var r = "";
    for (var i = 0; i < c.length; i++) {
        numDecrypted = decryptNum(c[i], pubK, privK);
        r = r + getLbyN(numDecrypted);
    }
    return r;
}

function encryptNum(m, pubK) {
    m = bignum(m);
    gM = pubK.g.pow(m);
    r = bignum(rand(0, pubK.n.toNumber()));
    rN = r.pow(pubK.n);
    gMrN = gM.mul(rN);
    n2= pubK.n.pow(2);
    c = gMrN.mod(n2);
    return c.toNumber();
}

function decryptNum(c, pubK, privK) {
    c = bignum(c);
    cLambda = c.pow(privK.lambda);
    n2 = pubK.n.pow(2);
    u = cLambda.mod(n2);
    lFunc = L(u, pubK.n);
    LMu = lFunc.mul(privK.mu);
    m = LMu.mod(pubK.n);
    return m;
}

function getLbyN(n) {
    for (var i = 0; i < numbers.length; i++) {
        if (numbers[i] == n) {
            return letters[i];
        }
    }
}

function getNbyL(l) {
    for (var i = 0; i < letters.length; i++) {
        if (letters[i] == l) {
            return numbers[i];
        }
    }
}

function rand(min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
}

function L(u, n) {
    u1 = u.sub(bignum(1));
    l = u1.div(n);
    return (l);
}

function gcd(a, b) {
    if (!b) {
        return a;
    }

    return gcd(b, a % b);
}

function lcm(a, b) {
    return ((a * b) / gcd(a, b));
}

function random(bitLength) {

    var wordLength = bitLength / 4 / 8;

    var randomWords = sjcl.random.randomWords(wordLength),
        randomHex = sjcl.codec.hex.fromBits(randomWords);

    return new BigInteger(randomHex, 16);

}


module.exports = {
    generateKeys : generatePaillierKeyPair,
    encrypt: encrypt,
    decrypt: decrypt
};

File: go/paillier.go
package main

import (
	"fmt"
	"math/big"
	"math/rand"
	"time"
)

type PaillierPublicKey struct {
	N *big.Int `json:"n"`
	G *big.Int `json:"g"`
}
type PaillierPrivateKey struct {
	Lambda *big.Int `json:"lambda"`
	Mu     *big.Int `json:"mu"`
}

type Paillier struct {
	PubK  PaillierPublicKey
	PrivK PaillierPrivateKey
}

func (paillier Paillier) generateKeyPair() Paillier {

	rand.Seed(time.Now().Unix())
	p := big.NewInt(int64(randPrime(minPrime, maxPrime)))
	q := big.NewInt(int64(randPrime(minPrime, maxPrime)))
	fmt.Print("p: ")
	fmt.Println(p)
	fmt.Print("q: ")
	fmt.Println(q)
	pq := new(big.Int).Mul(p, q)
	p1q1 := big.NewInt((p.Int64() - 1) * (q.Int64() - 1))
	gcd := new(big.Int).GCD(nil, nil, pq, p1q1)
	fmt.Print("gcd comprovation: ")
	fmt.Println(gcd)

	n := new(big.Int).Mul(p, q)
	lambda := big.NewInt(int64(paillier.lcm(float64(p.Int64())-1, float64(q.Int64())-1)))
	fmt.Print("lambda: ")
	fmt.Println(lambda)

	//g generation
	alpha := big.NewInt(int64(randInt(0, int(n.Int64()))))
	beta := big.NewInt(int64(randInt(0, int(n.Int64()))))
	alphan := new(big.Int).Mul(alpha, n)
	alphan1 := new(big.Int).Add(alphan, big.NewInt(1))
	betaN := new(big.Int).Exp(beta, n, nil)
	ab := new(big.Int).Mul(alphan1, betaN)
	n2 := new(big.Int).Mul(n, n)
	g := new(big.Int).Mod(ab, n2)
	//in some Paillier implementations use this:
	//g = new(big.Int).Add(n, big.NewInt(1))
	fmt.Print("g: ")
	fmt.Println(g)

	paillier.PubK.N = n
	paillier.PubK.G = g

	//mu generation
	Glambda := new(big.Int).Exp(g, lambda, nil)
	u := new(big.Int).Mod(Glambda, n2)
	L := paillier.L(u, n)
	mu := new(big.Int).ModInverse(L, n)

	paillier.PrivK.Lambda = lambda
	paillier.PrivK.Mu = mu

	return paillier
}
func (paillier Paillier) lcm(a, b float64) float64 {
	r := (a * b) / float64(gcd(int(a), int(b)))
	return r

}
func (paillier Paillier) L(u *big.Int, n *big.Int) *big.Int {
	u1 := new(big.Int).Sub(u, big.NewInt(1))
	L := new(big.Int).Div(u1, n)
	return L
}
func (paillier Paillier) encrypt(m string, pubK PaillierPublicKey) []int {
	var c []int
	mBytes := []byte(m)
	fmt.Print("mBytes: ")
	fmt.Println(mBytes)
	for _, byte := range mBytes {
		c = append(c, paillier.encryptInt(int(byte), pubK))
	}
	return c
}
func (paillier Paillier) decrypt(c []int, pubK PaillierPublicKey, privK PaillierPrivateKey) string {
	var m string
	var mBytes []byte
	for _, indC := range c {
		mBytes = append(mBytes, byte(paillier.decryptInt(indC, pubK, privK)))
	}
	m = string(mBytes)
	return m
}
func (paillier Paillier) encryptInt(char int, pubK PaillierPublicKey) int {
	m := big.NewInt(int64(char))
	gM := new(big.Int).Exp(pubK.G, m, nil)
	r := big.NewInt(int64(randInt(0, int(pubK.N.Int64()))))
	rN := new(big.Int).Exp(r, pubK.N, nil)
	gMrN := new(big.Int).Mul(gM, rN)
	n2 := new(big.Int).Mul(pubK.N, pubK.N)
	c := new(big.Int).Mod(gMrN, n2)
	return int(c.Int64())
}
func (paillier Paillier) decryptInt(val int, pubK PaillierPublicKey, privK PaillierPrivateKey) int {
	c := big.NewInt(int64(val))
	cLambda := new(big.Int).Exp(c, privK.Lambda, nil)
	n2 := new(big.Int).Mul(pubK.N, pubK.N)
	u := new(big.Int).Mod(cLambda, n2)
	L := paillier.L(u, pubK.N)
	LMu := new(big.Int).Mul(L, privK.Mu)
	m := new(big.Int).Mod(LMu, pubK.N)
	return int(m.Int64())
}

/*func (paillier Paillier) homomorphicAddition(c1 []int, c2 []int, pubK PaillierPublicKey) []int {
	var r []int
	for i := 0; i < len(c1); i++ {
		c1BigInt := big.NewInt(int64(c1[i]))
		c2BigInt := big.NewInt(int64(c2[i]))
		c1c2 := new(big.Int).Mul(c1BigInt, c2BigInt)
		n2 := new(big.Int).Mul(pubK.N, pubK.N)
		d := new(big.Int).Mod(c1c2, n2)
		r = append(r, int(d.Int64()))
	}
	return r
}*/

func (paillier Paillier) homomorphicAddition(c1 int, c2 int, pubK PaillierPublicKey) int {
	c1BigInt := big.NewInt(int64(c1))
	c2BigInt := big.NewInt(int64(c2))
	c1c2 := new(big.Int).Mul(c1BigInt, c2BigInt)
	n2 := new(big.Int).Mul(pubK.N, pubK.N)
	d := new(big.Int).Mod(c1c2, n2)
	r := int(d.Int64())
	return r
}
